blocks/blockstore/multi.go

```diff
@@ -0,0 +1,158 @@
+package blockstore
+
+// A very simple multi-blockstore that analogous to a unionfs Put and
+// DeleteBlock only go to the first blockstore all others are
+// considered readonly.
+
+import (
+	//"errors"
+	"context"
+
+	blocks "github.com/ipfs/go-ipfs/blocks"
+	cid "gx/ipfs/QmXUuRadqDq5BuFWzVU6VuKaSjTcNm1gNCtLvvP1TJCW4z/go-cid"
+	dsq "gx/ipfs/QmbzuUusHqaLLoNTDEVLcSF6vZDHZDLPC7p4bztRvvkXxU/go-datastore/query"
+)
+
+type LocateInfo struct {
+	Prefix string
+	Error  error
+}
+
+type MultiBlockstore interface {
+	Blockstore
+	GCLocker
+	FirstMount() Blockstore
+	Mounts() []string
+	Mount(prefix string) Blockstore
+	Locate(*cid.Cid) []LocateInfo
+}
+
+type Mount struct {
+	Prefix string
+	Blocks Blockstore
+}
+
+func NewMultiBlockstore(mounts ...Mount) *multiblockstore {
+	return &multiblockstore{
+		mounts: mounts,
+	}
+}
+
+type multiblockstore struct {
+	mounts []Mount
+	gclocker
+}
+
+func (bs *multiblockstore) FirstMount() Blockstore {
+	return bs.mounts[0].Blocks
+}
+
+func (bs *multiblockstore) Mounts() []string {
+	mounts := make([]string, 0, len(bs.mounts))
+	for _, mnt := range bs.mounts {
+		mounts = append(mounts, mnt.Prefix)
+	}
+	return mounts
+}
+
+func (bs *multiblockstore) Mount(prefix string) Blockstore {
+	for _, m := range bs.mounts {
+		if m.Prefix == prefix {
+			return m.Blocks
+		}
+	}
+	return nil
+}
+
+func (bs *multiblockstore) DeleteBlock(key *cid.Cid) error {
+	return bs.mounts[0].Blocks.DeleteBlock(key)
+}
+
+func (bs *multiblockstore) Has(c *cid.Cid) (bool, error) {
+	var firstErr error
+	for _, m := range bs.mounts {
+		have, err := m.Blocks.Has(c)
+		if have && err == nil {
+			return have, nil
+		}
+		if err != nil && firstErr == nil {
+			firstErr = err
+		}
+	}
+	return false, firstErr
+}
+
+func (bs *multiblockstore) Get(c *cid.Cid) (blocks.Block, error) {
+	var firstErr error
+	for _, m := range bs.mounts {
+		blk, err := m.Blocks.Get(c)
+		if err == nil {
+			return blk, nil
+		}
+		if firstErr == nil || firstErr == ErrNotFound {
+			firstErr = err
+		}
+	}
+	return nil, firstErr
+}
+
+func (bs *multiblockstore) Locate(c *cid.Cid) []LocateInfo {
+	res := make([]LocateInfo, 0, len(bs.mounts))
+	for _, m := range bs.mounts {
+		_, err := m.Blocks.Get(c)
```

Comment:
I am fine with that.  Not 100% happy with it as filestore Has doesn't really guarantee the block is available (and changing it so it does will make the Has() call really expensive), but its not a major problem.